### ISSUE TITLE
Http/2 client-side support

### ISSUE THREAD

**Issue by mimani** [akka/akka-http#1883]
Is there any plan to provide HTTP/2 client side support?

I am not sure if this is asked earlier as this seems quite common but could not found when I [searched](https://github.com/akka/akka-http/search?p=1&q=client+http+2&type=Issues&utf8=%E2%9C%93) it. I see there is [server side](https://github.com/akka/akka-http/issues/244) implementation but client side was out of scope that time.

**Comment by jrudolph** [comment]
It seems we haven't tracked it so far indeed. We are planning to provide a client implementation at some point. Nothing definite but it might be later this year.

**Comment by mushtaq** [comment]
@jrudolph is this going to be part of 10.2?

**Comment by raboof** [comment]
No, we expect to release 10.2.0 earlier than implementing HTTP/2 client-side support

**Comment by mtraul** [comment]
Is work for this being done already ? If not, I would love to contribute. Require Http/2 support for Apple APNS support

**Comment by raboof** [comment]
> Is work for this being done already ? If not, I would love to contribute.

No, we haven't started on this - it would be great if you could contribute! While we hope a lot of the work on the server-side implementation can be re-used when implementing the client side, this might still be a non-trivial feature: for example, since HTTP/2 allows multiplexing many parallel requests over a single TCP connection that introduces some interesting opportunities - and challenges!

> Require Http/2 support for Apple APNS support

:+1:

**Comment by jrudolph** [comment]
I had a quick try at this today in #3166. The goal for it would be to be able to get some base for a potential implementation into the code base that would help with contributions. Maybe you'd like to have a look at that and play with it, @mtraul? There's a big list of things that still need to be done, maybe I can finish it to a point where it already makes sense to merge it and then we can ticketify the list of remaining issues.

**Comment by pete-proton** [comment]
thanks, @jrudolph ! are we any closer after this PR https://github.com/akka/akka-http/pull/3166 has been merged?

**Comment by jrudolph** [comment]
Yes, we are closer but not yet far enough to announce something more publicly (won't happen for 10.2.0 but maybe later in the series, we still need to figure out where to put it in our internal roadmap).

The client-side is now somewhat working but no public API is available yet. If you want to play with it, you can use the internal API. https://github.com/jrudolph/akka-http/blob/cd5e40b2b555e5ef6195692dec7c8d67be9464a0/akka-http2-support/src/test/scala/akka/http/scaladsl/Http2ClientApp.scala is a working example that should work with 10.2.0-RC1.

**Comment by jrudolph** [comment]
Also, watch open tickets at https://github.com/akka/akka-http/labels/t%3Ahttp2 for potential outstanding issues.

**Comment by pete-proton** [comment]
thank you very much, @jrudolph. checking.
